File: violetadieu/Week_05/BOJ_2632_피자판매.py
from collections import defaultdict
import sys

pizza = int(input())
input = sys.stdin.readline
A, B = map(int, input().split())

PA = list()
PB = list()

answer = 0

CA = defaultdict(int)
CB = defaultdict(int)

for i in range(0, A):
    tmp = int(input())
    PA.append(tmp)

for i in range(0, B):
    tmp = int(input())
    PB.append(tmp)

CA[0] = 1
CB[0] = 1
CA[sum(PA)] = 1
CB[sum(PB)] = 1

# A 경우의 수
for i in range(A):
    total = PA[i]
    CA[total]+=1
    for p in range(1, A - 1):
        total += PA[(i + p) % A]
        CA[total]+=1

# B 경우의 수
for i in range(B):
    total = PB[i]
    CB[total]+=1
    for p in range(1, B - 1):
        total += PB[(i + p) % B]
        CB[total]+=1

for num in CA:
    if pizza - num in CB:
        answer += CA[num] * CB[pizza - num]

print(answer)

File: u1qns/Week_05/BOJ_2632_피자판매.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	
	static StringBuilder sb;
	static StringTokenizer st;
	static BufferedReader br;

	static int A[], B[];
	static int K, N, M, answer = 0;
	
	static int[] pick(final int pizza[])
	{
		// 피자의 모든 조합을 여기서 구한다 ! 
		int result[] = new int[K+1];
		
		int range = pizza.length;
		int sum = 0, cnt = 0;
		int total = 0;
		for(int i=0; i<range; ++i)
		{
			if(pizza[i] > K) continue;
			
			//System.out.printf("SUM : %d \n", pizza[i]);
			result[pizza[i]]++;
			total += pizza[i];
		}
		
		result[0] = 1;
		
		// 홀 피자 만드는 짓은 한번만 하고 싶다.
		
		for(int i=0; i<range; ++i)
		{
			sum = pizza[i];
			cnt = range-1;
			for(int j=i+1; cnt>0; ++j)
			{
				j = j%range;
				sum += pizza[j];
				if(sum > K) break;
				--cnt;
				//System.out.printf("SUM : %d \n", sum);
				result[sum]++;
			}
		}
		
		if(total <= K)
			result[total] = 1;
		
		return result;
	}
	
	static void solve()
	{
		// 1. A에서 가능한 조합을 만들자. 
		// 이 조합의 합이 K 이하라면 
		// 2. 나머지를 B에서 만들자
		
		int[] sumA = pick(A);
		//System.out.println("\n=====================");
		int[] sumB = pick(B);
		for(int i=0; i<=K; ++i)	
		{
			answer += (sumA[i] * sumB[K-i]);
			//System.out.printf("answer += %d * %d = %d\n", sumA[i], sumB[K-i], answer);

		}
	}
	
	public static void main(String[] args) throws IOException {
	
		br = new BufferedReader(new InputStreamReader(System.in));

		K = Integer.parseInt(br.readLine());

		st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		
		A = new int[N]; B = new int[M];
		for(int i=0; i<N; ++i)
			A[i] = Integer.parseInt(br.readLine());
		
		for(int i=0; i<M; ++i)
			B[i] = Integer.parseInt(br.readLine());
		
		
		solve();
		
		System.out.print(answer);
		
	} // main

	
}
